File: src/main/java/eu/mshade/enderchest/marshal/metadata/MetadataKeyValueBinaryTagMarshal.java
package eu.mshade.enderchest.marshal.metadata;

import eu.mshade.enderframe.metadata.MetadataKey;
import eu.mshade.enderframe.metadata.MetadataKeyValue;
import eu.mshade.enderframe.metadata.MetadataKeyValueBucket;
import eu.mshade.enderframe.metadata.world.WorldMetadataType;
import eu.mshade.enderframe.world.Difficulty;
import eu.mshade.enderframe.world.Dimension;
import eu.mshade.enderframe.world.LevelType;
import eu.mshade.enderframe.world.metadata.*;
import eu.mshade.mwork.binarytag.BinaryTag;
import eu.mshade.mwork.binarytag.BinaryTagDriver;
import eu.mshade.mwork.binarytag.BinaryTagDynamicMarshal;
import eu.mshade.mwork.binarytag.entity.CompoundBinaryTag;
import eu.mshade.mwork.binarytag.entity.LongBinaryTag;
import eu.mshade.mwork.binarytag.entity.StringBinaryTag;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

public class MetadataKeyValueBinaryTagMarshal implements BinaryTagDynamicMarshal {

    private static Logger LOGGER = LoggerFactory.getLogger(MetadataKeyValueBinaryTagMarshal.class);

    private final Map<String, MetadataKey> metadataKeyByName = new HashMap<>();
    private final Map<MetadataKey, Function<MetadataKeyValue<?>, BinaryTag<?>>> metadataKeyValueToBinaryTag = new HashMap<>();
    private final Map<MetadataKey, Function<BinaryTag<?>, MetadataKeyValue<?>>> binaryTagToMetadataKeyValue = new HashMap<>();

    public MetadataKeyValueBinaryTagMarshal(BinaryTagDriver binaryTagDriver) {
        this.register(WorldMetadataType.NAME, metadataKeyValue -> new StringBinaryTag((String) metadataKeyValue.getMetadataValue()), binaryTag -> new NameWorldMetadata((String) binaryTag.getValue()));
        this.register(WorldMetadataType.SEED, metadataKeyValue -> new LongBinaryTag((Long) metadataKeyValue.getMetadataValue()), binaryTag -> new SeedWorldMetadata((Long) binaryTag.getValue()));
        this.register(WorldMetadataType.DIMENSION, metadataKeyValue -> binaryTagDriver.marshal(metadataKeyValue.getMetadataValue()), binaryTag -> new DimensionWorldMetadata(binaryTagDriver.unMarshal(binaryTag, Dimension.class)));
        this.register(WorldMetadataType.LEVEL_TYPE, metadataKeyValue -> binaryTagDriver.marshal(metadataKeyValue.getMetadataValue()), binaryTag -> new LevelTypeWorldMetadata(binaryTagDriver.unMarshal(binaryTag, LevelType.class)));
        this.register(WorldMetadataType.DIFFICULTY, metadataKeyValue -> binaryTagDriver.marshal(metadataKeyValue.getMetadataValue()), binaryTag -> new DifficultyWorldMetadata(binaryTagDriver.unMarshal(binaryTag, Difficulty.class)));
        LOGGER.info("Register {} metadataKeyValues", metadataKeyByName.size());
    }

    public CompoundBinaryTag serialize(MetadataKeyValueBucket metadataKeyValueBucket){
        CompoundBinaryTag compoundBinaryTag = new CompoundBinaryTag();

        for (MetadataKeyValue<?> metadataKeyValue : metadataKeyValueBucket.getMetadataKeyValues()) {
            compoundBinaryTag.putBinaryTag(metadataKeyValue.getMetadataKey().getName(), metadataKeyValueToBinaryTag.get(metadataKeyValue.getMetadataKey()).apply(metadataKeyValue));
        }

        return compoundBinaryTag;
    }

    public MetadataKeyValueBucket deserialize(CompoundBinaryTag compoundBinaryTag){
        MetadataKeyValueBucket metadataKeyValueBucket = new MetadataKeyValueBucket();

        compoundBinaryTag.getValue().forEach((s, binaryTag) -> {
            metadataKeyValueBucket.setMetadataKeyValue(binaryTagToMetadataKeyValue.get(metadataKeyByName.get(s)).apply(binaryTag));
        });

        return metadataKeyValueBucket;
    }

    public <K extends MetadataKey> void register(K metadataKey, Function<MetadataKeyValue<?>, BinaryTag<?>> metadataKeyValueToBinaryTag, Function<BinaryTag<?>, MetadataKeyValue<?>> binaryTagToMetadataKeyValue){
        this.metadataKeyByName.put(metadataKey.getName(), metadataKey);
        this.metadataKeyValueToBinaryTag.put(metadataKey, metadataKeyValueToBinaryTag);
        this.binaryTagToMetadataKeyValue.put(metadataKey, binaryTagToMetadataKeyValue);
        LOGGER.debug("Register "+metadataKey);
    }

}

File: src/main/java/eu/mshade/enderchest/marshal/metadata/MetadataKeyValueBinaryTagMarshal.kt
package eu.mshade.enderchest.marshal.metadata

import eu.mshade.enderframe.metadata.MetadataKey
import eu.mshade.enderframe.metadata.MetadataKeyValueBucket
import eu.mshade.enderframe.metadata.world.WorldMetadataType
import eu.mshade.enderframe.world.block.BlockMetadataType.AXIS
import eu.mshade.enderframe.world.block.BlockMetadataType.CHECK_DECAY
import eu.mshade.enderframe.world.block.BlockMetadataType.DECAYABLE
import eu.mshade.enderframe.world.block.BlockMetadataType.EXTRA
import eu.mshade.enderframe.world.block.BlockMetadataType.FACE
import eu.mshade.enderframe.world.block.BlockMetadataType.HALF
import eu.mshade.enderframe.world.block.BlockMetadataType.POWER
import eu.mshade.enderframe.world.block.BlockMetadataType.POWERED
import eu.mshade.enderframe.world.block.BlockMetadataType.SEAMLESS
import eu.mshade.enderframe.world.block.BlockMetadataType.SHAPE
import eu.mshade.mwork.binarytag.BinaryTag
import eu.mshade.mwork.binarytag.BinaryTagDriver
import eu.mshade.mwork.binarytag.BinaryTagDynamicMarshal
import eu.mshade.mwork.binarytag.entity.CompoundBinaryTag
import org.slf4j.LoggerFactory

class MetadataKeyValueBinaryTagMarshal(binaryTagDriver: BinaryTagDriver?) : BinaryTagDynamicMarshal {
    private val metadataKeyByName: MutableMap<String, MetadataKey> = HashMap()
    private val metadataKeyValueBuffer: MutableMap<MetadataKey?, MetadataKeyValueBuffer> = HashMap()

    init {
        register(WorldMetadataType.NAME, NameWorldMetadataBuffer())
        register(WorldMetadataType.SEED, SeedWorldMetadataBuffer())
        register(WorldMetadataType.DIMENSION, DimensionWorldMetadataBuffer(binaryTagDriver!!))
        register(WorldMetadataType.LEVEL_TYPE, LevelTypeWorldMetadataBuffer(binaryTagDriver))
        register(WorldMetadataType.DIFFICULTY, DifficultyWorldMetadataBuffer(binaryTagDriver))

        register(EXTRA, ExtraBlockMetadataBuffer())
        register(FACE, FaceBlockMetadataBuffer())
        register(HALF, HalfBlockMetadataBuffer())
        register(SHAPE, ShapeBlockMetadataBuffer())
        register(AXIS, AxisBlockMetadataBuffer())
        register(POWERED, PoweredBlockMetadataBuffer())
        register(POWER, PowerBlockMetadataBuffer())
        register(DECAYABLE, DecayableBlockMetadataBuffer())
        register(CHECK_DECAY, CheckDecayBlockMetadataBuffer())
        register(SEAMLESS, SeamlessBlockMetadataBuffer())

        LOGGER.info("Register {} metadataKeyValueBuffers", metadataKeyByName.size)
    }

    fun serialize(metadataKeyValueBucket: MetadataKeyValueBucket): CompoundBinaryTag {
        val compoundBinaryTag = CompoundBinaryTag()
        for (metadataKeyValue in metadataKeyValueBucket.metadataKeyValues) {
            compoundBinaryTag.putBinaryTag(
                metadataKeyValue.metadataKey.name, metadataKeyValueBuffer[metadataKeyValue.metadataKey]!!
                    .write(metadataKeyValue)
            )
        }
        return compoundBinaryTag
    }

    fun deserialize(compoundBinaryTag: CompoundBinaryTag): MetadataKeyValueBucket {
        val metadataKeyValueBucket = MetadataKeyValueBucket()
        compoundBinaryTag.value.forEach { (s: String, binaryTag: BinaryTag<*>?) ->
            val metadataKeyValue = metadataKeyValueBuffer[metadataKeyByName[s]]!!
                .read(binaryTag!!)
            metadataKeyValueBucket.setMetadataKeyValue(metadataKeyValue)
        }
        return metadataKeyValueBucket
    }

    fun <K : MetadataKey?> register(metadataKey: K, metadataKeyValueBuffer: MetadataKeyValueBuffer) {
        metadataKeyByName[metadataKey!!.name] = metadataKey
        this.metadataKeyValueBuffer[metadataKey] = metadataKeyValueBuffer
        LOGGER.debug("Register $metadataKey")
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(MetadataKeyValueBinaryTagMarshal::class.java)
    }
}